### ISSUE TITLE
Invalid session key returned

### ISSUE THREAD

**Issue by jmccabe** [axnsantana/python-limesurvey-rc#1]
Calling get_session_key worked fine, but the next call to export_responses returned an invalid session key error. I haven't investigated this further yet but noticed, in __format_params, there's the line:

params['sSessionKey']=self.session_key

Is the 'sSessionKey' string correct there?

**Comment by axnsantana** [comment]
Hi,

Yes, it is the name of the param inside the LimeSurvey API:
http://api.limesurvey.org/classes/remotecontrol_handle.html#method_export_responses
You should call get_session_key before call export_responses.

You can use the call_rpc method to check the entire response from LimeSurvey.
This __format_params is used in all other methods, you can check it with a simpler method like: list_surveys.
Other point to consider is if the user has permission to access the responses.

If you provide more information, maybe I can help you more.

**Comment by jmccabe** [comment]
Many thanks for your reply. I'll try some more debugging later, but get_summary also failed and the user name and password I'm using to get the session key are definitely correct. I've used wireshark to look at the data being sent on the wire and it's understanding the json ok.  At this point, the only thing I can think of is that the server is being particular about the order of the parameters (the session key is being sent after the survey id). I'll get back in touch if I get any further, or not. 

Thanks again for your help, and making this available! 

John

**Comment by axnsantana** [comment]
Okay, any news please let me know!

I am without access to my LimeSurvey server, in few days I will return and make some tests with the methods get_summary and export_responses.
But the first thing to check is the return of get_session_id. 
It should be an alphanumeric sequence. 
You can check it with the entry saved in the LimeSurvey Database.

The API calls this method before any action. It deletes all expired keys and checks if yours is one of the entries. 

```
protected function _checkSessionKey($sSessionKey)
    {
        $criteria = new CDbCriteria;
        $criteria->condition = 'expire < ' . time();
        Session::model()->deleteAll($criteria);
        $oResult = Session::model()->findByPk($sSessionKey);

        if (is_null($oResult))
            return false;
        else
        {
            $this->_jumpStartSession($oResult->data);
            return true;
        }
    }
```

To test if the problem is the parameters order, you can check a basic method without any parameter, like: list_surveys, list_users.

**Comment by jmccabe** [comment]
Alex, I wonder if this is relevant - https://bugs.limesurvey.org/view.php?id=8314. I have done some more debugging. I've been able to force the order, albeit a bit randomly, and some functions work. Others still don't appear happy. If you can test your code with a recent limesurvey it would be useful. I tried an alternative python module from github.com and found similar problems.

**Comment by jmccabe** [comment]
More news...

I'm fairly convinced now that order is important.

I tried working with the code here: https://github.com/psychopenguin/limesurvey, and expanding bit by bit.

I've reviewed various bits and pieces on the web, including https://manual.limesurvey.org/RemoteControl_2_API#Python_example_and_glue, https://github.com/LimeSurvey/LimeSurvey/pull/147 and https://www.limesurvey.org/forum/can-i-do-this-with-limesurvey/104171-limesurvey-remotecontrol-api,-python-glue.

It seems that, unless limesurvey has been updated to use JSON-RPC2, all the parameters for the function calls need to be supplied in the order they're specified in the functions. Named parameters can be supplied, but they must still match the declaration order of the API.

Based on the "can-i-do-this-with-limesurvey' link above I modified your __format_params() function to return a list rather than a dictionary as follows:

```
    def __format_params(self, params):
        del params['self']
        rparams = []
        if self.session_key is not None:
            rparams = [ self.session_key ]
        rparams = rparams + [params[k] for k,v in params.iteritems() if v is not None]
        return rparams

```

This meant that the parameters are passed in the json request as an array. However the problem with this approach is that locals().copy(), which is used to pass params into __format_params(), doesn't necessarily provide the parameters in the order they're specified in the function declaration. In particular, I tried def get_summary(self, iSurveyID, sStatName=None), which was still failing, in this case with an Invalid surveyid error. When I print out the contents of locals().copy() in the get_summary() function it returns:

{'sStatName': 'all', 'self': <pylimerc.PyLimeRc instance at 0x01564B48>, 'iSurveyID': 111828}

In other words, when __format_params() iterates through that dictionary it adds sStatName before iSurveyID.

This seems rather unfortunate. If you get a chance though to check this is true please let me know.

**Comment by axnsantana** [comment]
John, I created a new branch. This version puts the params in order as a list.
I did some tests in my server with list_questions and get_summary, it was okay.
Please, test it in your case. Let me know about the results.

And thanks for report this problem and all research to solve it.
teste.py

```
a = PyLimeRc(url)
a.get_session_key('#####','@@@@@')
a.list_surveys()
a.list_questions('964288')
summary = a.get_summary('964288','all')
print summary
```

output

```
Method: get_session_key
Params before: {'username': '#####', 'password': '@@@@@'}
Params sorted: ['#####', '@@@@@']
Method: list_surveys
Params before: {'sSessionKey': u'7x3nic4urkeckmj3v93aawaix3nnyvuu'}
Params sorted: [u'7x3nic4urkeckmj3v93aawaix3nnyvuu']
Method: list_questions
Params before: {'sSessionKey': u'7x3nic4urkeckmj3v93aawaix3nnyvuu', 'iSurveyID': '964288'}
Params sorted: [u'7x3nic4urkeckmj3v93aawaix3nnyvuu', '964288']
Method: get_summary
Params before: {'sStatName': 'all', 'iSurveyID': '964288', 'sSessionKey': u'7x3nic4urkeckmj3v93aawaix3nnyvuu'}
Params sorted: [u'7x3nic4urkeckmj3v93aawaix3nnyvuu', '964288', 'all']
[]

```

**Comment by jmccabe** [comment]
Hi Alex,

Thanks for that. I'll try to take a look at it shortly but I don't currently have yaml installed in my Python stuff. For the moment, I'm passing the params for each function I'm interested in as a list instead of using the locals().copy() and that seems to work. As I said though, when I get a chance I'll try to look at this version.

John

**Comment by axnsantana** [comment]
Hi Jonh,
I am closing this issue and merging the branches.
Thanks for the contribution.

**Comment by jmccabe** [comment]
You're welcome. Sorry I never got round to trying it properly; it was the yaml stuff that held me back!
All the best
John
